File: Python/A_Project/B_Snake_and_Food/FoodItemFactory/food_spawner.py
from FoodItemFactory.food_factory import FoodFactory


class FoodSpawner:
    def __init__(self, positions, bonus_every_k: int):
        self.positions = positions
        self.bonus_every_k = max(2, bonus_every_k)
        self.index = 0

    def has_more(self) -> bool:
        return self.index < len(self.positions)

    def current_item(self):
        if not self.has_more():
            return None
        pos = self.positions[self.index]
        food_type = "bonus" if ((self.index + 1) % self.bonus_every_k) == 0 else "normal"
        return FoodFactory.create_food(pos, food_type)

    def consume(self):
        if self.has_more():
            self.index += 1

    def get_index(self) -> int:
        return self.index

File: Ruby/A_Project/B_Snake_And_Food/FoodItemFactory/food_spawner.rb
# frozen_string_literal: true

require_relative "food_factory"

class FoodSpawner
  def initialize(positions, bonus_every_k)
    unless positions.respond_to?(:empty?) && !positions.empty?
      raise ArgumentError, "positions must be a non-empty list or tuple"
    end

    @positions = positions
    @bonus_every_k = [bonus_every_k.to_i, 2].max
    @index = 0
  end

  def has_more
    @index < @positions.length
  end

  def current_item
    return nil unless has_more

    pos = @positions[@index]
    food_type = ((@index + 1) % @bonus_every_k).zero? ? "bonus" : "normal"
    FoodFactory.create_food(pos, food_type)
  end

  def consume
    @index += 1 if has_more
  end

  def get_index
    @index
  end
end
